Add support to set ciphers for TLSv1.3

mattheusv opened the pull request:
Currently on PgBouncer it's not possible to set ciphers for TLS v1.3. This is
because for TLS v1.3 we need to use a different API to set the ciphers, e.g:
```
client_tls_protocols=tlsv1.3
client_tls_ciphers=TLS_CHACHA20_POLY1305_SHA256
client_tls_sslmode=require
```

When connect using `psql` we get:
`SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, compression: off, ALPN: none)`

This PR change the `tls_config_set_ciphers` to receive an optional ciphers_suites
parameter to set on `tls_config` so that it can be used later on
`tls_configure_ssl` to call the new OpenSSL API to set the v1.3 ciphers.

Once we get this committed, I'll update the PgBouncer code to add a config for
the TLS v1.3 ciphers to pass through the tls_config_set_ciphers function.

---
Some open questions:

On `tls_config_set_ciphers` I'm only setting the cipher_suites on tls_config,
and I'm not sure if I also should call the `SSL_CTX_set_ciphersuites` for TLS
v1.3 as this function also call the `SSL_CTX_set_cipher_list` for TLS v1.2 or
older.

I've hard-coded the priority parameter value from `SSL_get_cipher_list` on
usual/tls/tls_util.c and I'm not sure if we need to make this configurable or
use another value.

mattheusv committed:
Add support to set ciphers for TLSv1.3

paulwouters commented:
do NOT call SSL_CTX_set_ciphersuites for any default configurations. Do not manually set what is the default configuration either in libusual or pgbouncer, so that the crypto library default, steered by the system wide crypto policies, actually takes effect. Otherwise we are back at the same problems we ended up fixing for TLS 1.2 where 10 year old defaults were being set through the call chain.

I'm confused why you only got TLS_AES_256_GCM_SHA384. It is expected that you get both AES-GCM and ChaCha20-Poly1305 on a default TLS 1.3 ciphersuite setting. Perhaps a bug is preventing this that needs solving instead of this PR.

mattheusv commented:
> do NOT call SSL_CTX_set_ciphersuites for any default configurations.
> Do not manually set what is the default configuration either in
> libusual or pgbouncer, so that the crypto library default, steered by
> the system wide crypto policies, actually takes effect. Otherwise we
> are back at the same problems we ended up fixing for TLS 1.2 where 10
> year old defaults were being set through the call chain.
> 
Do we want to support the "default", "secure", "normal" and "fast"
values for ciphers suites (v1.3) as for cipher list (v1.2)? I think that
we can create another function `tls_config_set_ciphers_v13` that works
very similar with the current `tls_config_set_ciphers`, then the pg
bouncer decides what function it should call based on the user
configuration, what do you think?

> I'm confused why you only got TLS_AES_256_GCM_SHA384. It is expected
> that you get both AES-GCM and ChaCha20-Poly1305 on a default TLS 1.3
> ciphersuite setting. Perhaps a bug is preventing this that needs
> solving instead of this PR.
>
This is something that need to be fixed before this PR?

paulwouters commented:
No, the concepts of default, secure, normal and fast are completely outdated, and their settings in the last 10 years were very wrong. I'd be happy to create a PR that removes them entirely. See https://github.com/libusual/libusual/pull/40

Again, the important thing is not to call SSL functions to declare a default cipher, but to let the library pick its buildin one, unless the user (not the application!) configures a specific setting.

I think fixing whatever bug you are seeing would make this PR unnecessary. I'd rather rip out more of the old code of groupnames instead of fixing it for TLS 1.3 with other hardcoding or passing along default config settings that override the system library defaults.

mattheusv commented:
> Again, the important thing is not to call SSL functions to declare a default cipher, but to let the library pick its buildin one, unless the user (not the application!) configures a specific setting.
> 
Please, take a look at the new version. I've added an explicit check that only
set the cipher suites if the protocol version is 1.3 and also added an explicit
validation to only set the if value is not NULL and not empty (The set_string
function also check if the value is NULL before setting). I've added the same
validation on tls_configure_tls function.

On PgBouncer the cipher_suites will be NULL if the user don't specify a one, so
it will use the default value from the library.

With this commit, if the user only set `client_tls_protocols=tlsv1.3` on
pg_bouncer.ini we will not set anything and it will use the library default. To
set a custom cipher for TLS v1.3 the user will need to set a new config value
like `client_tls13_ciphers=TLS_CHACHA20_POLY1305_SHA256` so that it will be
pased to `tls_config_set_ciphers` function.

> I think fixing whatever bug you are seeing would make this PR unnecessary. I'd rather rip out more of the old code of group names instead of fixing it for TLS 1.3 with other hard coding or passing along default config settings that override the system library defaults.
>
Sorry, but I'm not sure if I understand your point here. The
`TLS_CHACHA20_POLY1305_SHA256` cipher is not being set because we need to use a
different OpenSSL API to set the cipher for v1.3 which is the goal of this PR.

--- 
(I'm still not sure if you generally agree with the idea proposed in this PR)

paulwouters reviewed (approved):
Thanks for the discussion and confirmation you don't set anything by default.

paulwouters commented:
(I'll do a PR to remove the old bad keywords and see if it will get accepted)

mattheusv commented:
> Thanks for the discussion and confirmation you don't set anything by default.

Thanks for reviewing!

petere commented:
@mattheusv Do you have a corresponding pull request for pgbouncer/pgbouncer?

petere commented:
pgbouncer no longer uses this copy of libusual (see pgbouncer/pgbouncer#1331). If you want to have your code end up in pgbouncer, please make a pull request against the pgbouncer/pgbouncer repository. (And you can close this one.)

mattheusv commented:
Closing this PR to open a new one on pgbouncer

mattheusv commented:
> @mattheusv Do you have a corresponding pull request for pgbouncer/pgbouncer?

Sorry for delay. I already have the necessary changes for pgbouncer locally, but I didn't open a PR on pgbouncer yet. I'll open a new one with all the changes for pgbouncer and libusual
